### PR TITLE
SQL: Import Tableau connector

### PR DESCRIPTION
**Pull request by bpintea**
Initial Tableau connector import.

### CODE REVIEW

**Review by astefan — APPROVED** [review]
LGTM. I guess a follow up issue (and PR) is needed for adding Logstash data loading into Tableau.

File: x-pack/plugin/sql/connectors/tableau/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+The Tableau Connector works in tandem with the Elastic JDBC driver to facilitate the query of Elasticsearch. It gives users a simple way to query Elasticsearch data from Tableau.
+After providing basic connection and authentication information, users can easily select Elasticsearch indices for use in Tableau Desktop and Tableau Server.
+
+## Installation
+
+1. Tableau connector for Elasticsearch installation:
+ - Go to the [Connector Download](https://www.elastic.co/downloads/tableau-connector) page.
+ - Download the _.taco_ connector file.
+ - Move the _.taco_ file here:
+    - Windows: C:\Users[Windows User]\Documents\My Tableau Repository\Connectors
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/connectors/tableau/README.md:12]: Fixed.

**Comment by bpintea** [comment]
Thanks, @astefan.

> I guess a follow up issue (and PR) is needed for adding Logstash data loading into Tableau.

I've added a script to automise running the TDVT tools and added a paragraph on it, so I've added the steps for Logstash to this PR too. Let me know please if you'd like any changes to those or if it can go in as is.

**Comment by bpintea** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/connectors/tableau/README.md (hunk 1 above)
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/README.md:12]:
```suggestion
    - Windows: C:\Users\[Windows User]\Documents\My Tableau Repository\Connectors
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+1. Create the Elasticsearch indices `calcs` and `staples` and use for them the [mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html) under these links: [calcs](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L27) and [stapes](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L87). Create an ingest [pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/put-pipeline-api.html) called `calcs-pipeline` with the definition [here](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L62).
+
+2. Adapt the [config file](https://www.elastic.co/guide/en/logstash/current/configuration-file-structure.html#configuration-file-structure) under the `logstash` folder, updating the <path>, <host> and <password> tags in it.
+
+3. Relaunch Logstash and wait until the files have been ingested. `calcs` index will need to have 17 documents and `staples` 54860.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:57]:
```suggestion
3. Relaunch Logstash using the updated config file in previous step and wait until the files have been ingested. `calcs` index will need to have 17 documents and `staples` 54860.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+<TODO: detailed sequence of steps>
+
+### 3. Import using Elastic's Logstash application
+
+[Logstash](https://www.elastic.co/guide/en/logstash/current/running-logstash-windows.html) is a Java application that can ingest data from various sources and in various formats and upload them into an Elasticsearch instance.
+
+Note that Logstash will add a few extra fields per row ("document" in Elasticsearch lingo); these shouldn't interfere with the testing, however.
+
+0. Download TDVT's data files [Calcs.csv](https://raw.githubusercontent.com/tableau/connector-plugin-sdk/tdvt-2.1.9/tests/datasets/TestV1/Calcs.csv) and [Staples_utf8.csv](https://raw.githubusercontent.com/tableau/connector-plugin-sdk/tdvt-2.1.9/tests/datasets/TestV1/Staples_utf8.csv) and place them into a directory that will be reachable by Logstash.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:51]:
```suggestion
1. Download TDVT's data files [Calcs.csv](https://raw.githubusercontent.com/tableau/connector-plugin-sdk/tdvt-2.1.9/tests/datasets/TestV1/Calcs.csv) and [Staples_utf8.csv](https://raw.githubusercontent.com/tableau/connector-plugin-sdk/tdvt-2.1.9/tests/datasets/TestV1/Staples_utf8.csv) and place them into a directory that will be reachable by Logstash.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+1. Create the Elasticsearch indices `calcs` and `staples` and use for them the [mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html) under these links: [calcs](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L27) and [stapes](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L87). Create an ingest [pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/put-pipeline-api.html) called `calcs-pipeline` with the definition [here](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L62).
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:53]:
```suggestion
2. Create the Elasticsearch indices `calcs` and `staples` and use for them the [mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html) under these links: [calcs](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L27) and [stapes](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L87). Create an ingest [pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/put-pipeline-api.html) called `calcs-pipeline` with the definition [here](https://github.com/elastic/elasticsearch-sql-odbc/blob/577cd2fa1ed257e42081a082682c8c089b179565/test/integration/data.py#L62).
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+2. Adapt the [config file](https://www.elastic.co/guide/en/logstash/current/configuration-file-structure.html#configuration-file-structure) under the `logstash` folder, updating the <path>, <host> and <password> tags in it.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:55]:
```suggestion
3. Adapt the [config file](https://www.elastic.co/guide/en/logstash/current/configuration-file-structure.html#configuration-file-structure) under the `logstash` folder, updating the <path>, <host> and <password> tags in it.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+3. Relaunch Logstash using the updated config file in previous step and wait until the files have been ingested. `calcs` index will need to have 17 documents and `staples` 54860.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:57]:
```suggestion
4. Relaunch Logstash using the updated config file in previous step and wait until the files have been ingested. `calcs` index will need to have 17 documents and `staples` 54860.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+
+## Running TDVT
+
+### Automated
+
+0. Place Elasticsearch JDBC driver into [Tableau's driver folder](https://help.tableau.com/current/pro/desktop/en-us/examples_otherdatabases_jdbc.htm) under: `C:\Program Files\Tableau\Drivers`.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:64]:
```suggestion
1. Place Elasticsearch JDBC driver into [Tableau's driver folder](https://help.tableau.com/current/pro/desktop/en-us/examples_otherdatabases_jdbc.htm) under: `C:\Program Files\Tableau\Drivers`.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+1. Place the `.taco` file either in Tableau's dedicated connectors directory, `C:\Users\[Windows User]\Documents\My Tableau Repository\Connectors`, or a custom one ("<taco dir path>").
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:66]:
```suggestion
2. Place the `.taco` file either in Tableau's dedicated connectors directory, `C:\Users\[Windows User]\Documents\My Tableau Repository\Connectors`, or a custom one ("<taco dir path>").
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+2. Use the `tdvt_run.py` application, that clone the TDVT SDK repo, setup config files and launch the TDVT run:
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:68]:
```suggestion
3. Use the `tdvt_run.py` application, that clone the TDVT SDK repo, setup config files and launch the TDVT run:
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    ```
+    python3 ./tdvt_run.py -u "http://user:pass@elastic-host:9200" -t <taco dir path>
+    ```
+
+### Manually
+
+Setting up the TDVT testing involves following the steps detailed in the [official documentation](https://tableau.github.io/connector-plugin-sdk/docs/tdvt). The "fragment" in parantheses reference the respective chapters in the documentation. It is recommended to execute each test run starting afresh.
+
+0. Same as in the automated testing.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:77]:
```suggestion
1. Same as in the automated testing.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+1. Create new Tableau data sources for the `calcs` and `Staple` tables (#`Test a new data source`), or, alternatively, use those available already in this repo.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:79]:
```suggestion
2. Create new Tableau data sources for the `calcs` and `Staple` tables (#`Test a new data source`), or, alternatively, use those available already in this repo.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+	```
+	.\tableau.exe -DConnectPluginsPath=<path> -DDisableVerifyConnectorPluginSignature=true
+	```
+	where `<path>` is either the path to the directory containing the `.taco` connector *or* the path to the directory containing the connector directory, if this isn't yet packaged.
+  
+	**Note**: When connecting, make sure you pass the `timezone=Z` parameter into the `Additional settings` field of the connection dialog. This sets the timezone for the time data to UTC; if this isn't set, the JDBC driver will use JVM's/system's time zone, which will then result in some failed tests if the machine's not set to the UTC timezone.
+
+	Save the TDS files as `cast_calcs.elastic.tds` and `Staples.elastic.tds`
+
+2. Setup a TDVT "workspace" (#`Set up`), i.e. a directory containing the test files.
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:90]:
```suggestion
3. Setup a TDVT "workspace" (#`Set up`), i.e. a directory containing the test files.
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+	Either package TDVT and install it as a Python PIP module (recommended, if [working](https://github.com/tableau/connector-plugin-sdk/issues/534)), or simply copy the `tdvt` directory of the repo into the "workspace" directory. Invoking the TDVT will then be done as `py -3 -m tdvt.tdvt <params>`, or `py -3 .\tdvt\tdvt_launcher.py <params>`, respectively. In the steps below the invokation will be indicated by the `$TDVT` call.
+	```
+	$TDVT action --setup
+	```
+
+	Copy/move the above saved `*.tds` files into the just created `tds` directory in the workspace.
+	Edit `config/tdvt/tdvt_override.ini` to update `TAB_CLI_EXE_X64` definition.
+
+3. Generate the tests by invoking TDVT as follows:
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:99]:
```suggestion
4. Generate the tests by invoking TDVT as follows:
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	```
+	$TDVT action --add_ds elastic
+	```
+
+	When asked for a password, use the same value as in step "2." above when connecting to Elasticsearch.
+	For the "logical query config" use `simple_lower`.
+
+	Edit the `elastic.ini` file in the `config` directory in the workspace and add the following line under the `[Datasource]` section: `CommandLineOverride = -DConnectPluginsPath=<path> -DDisableVerifyConnectorPluginSignature=true`, where `<path>` has the same value as in step "1.".
+
+4. Run the tests:
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:110]:
```suggestion
5. Run the tests:
```

File: x-pack/plugin/sql/connectors/tableau/tdvt/README.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+4. Generate the tests by invoking TDVT as follows:
+
+	```
+	$TDVT action --add_ds elastic
+	```
+
+	When asked for a password, use the same value as in step "2." above when connecting to Elasticsearch.
+	For the "logical query config" use `simple_lower`.
+
+	Edit the `elastic.ini` file in the `config` directory in the workspace and add the following line under the `[Datasource]` section: `CommandLineOverride = -DConnectPluginsPath=<path> -DDisableVerifyConnectorPluginSignature=true`, where `<path>` has the same value as in step "1.".
```
Comment (bpintea) [inline x-pack/plugin/sql/connectors/tableau/tdvt/README.md:108]:
```suggestion
	Edit the `elastic.ini` file in the `config` directory in the workspace and add the following line under the `[Datasource]` section: `CommandLineOverride = -DConnectPluginsPath=<path> -DDisableVerifyConnectorPluginSignature=true`, where `<path>` has the same value as in step "2.".
```

File: x-pack/plugin/sql/connectors/tableau/README.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+1. Tableau connector for Elasticsearch installation:
+ - Go to the [Connector Download](https://www.elastic.co/downloads/tableau-connector) page.
+ - Download the _.taco_ connector file.
+ - Move the _.taco_ file here:
+    - Windows: C:\Users\[Windows User]\Documents\My Tableau Repository\Connectors
+    -  macOS: /Users/[user]/Documents/My Tableau Repository/Connectors
+2. Elasticsearch JDBC Driver Installation:
+ - Go to the [Driver Download](https://www.elastic.co/downloads/jdbc-client) page.
+ - Download the Elasticsearch JDBC Driver _.jar_ file and move it into the following directory:
```
Comment (matriv) [inline x-pack/plugin/sql/connectors/tableau/README.md:16]: Should we mention something about appropriate versions based on the es node/cluster version?

Comment (bpintea) [reply x-pack/plugin/sql/connectors/tableau/README.md:16]: That's a good point. I would however follow up on this:
* once the connector releasing is hooked into the RM (and we'll then easily be able to pair the connector with the driver, version-wise);
* with a doc PR stating our drivers' compatibility policy that we can refer to (from here, among other places).

**Comment by bpintea** [comment]
@elasticmachine update branch

**Review by costin — COMMENTED** [review]
Some of the files in this PR seem to have originated in the Tableau Connector SDK.
I'm not sure whether there's some git magic that can be used to download these directly from their original repo but assuming there's none or it is too complicated, it's important to clarify the situation and, if any of their resources are used, clearly indicate the source, copyright and license (links are enough).

**Comment by bpintea** [comment]
@elasticmachine update branch

**Review by costin — APPROVED** [review]
I wasn't clear in my original notes so here goes my LGTM

**Pull request merged**